### DISCUSSION TITLE
Dashvis not working with Dash Pages?

### DISCUSSION THREAD

**Discussion by McGinn1014** [lewkoo/dashvis#12] [Q&A]
Hi! First I'd like to say that this project is wonderful and works incredibly well! I am creating a Dash App as my first 'side project' and found Visdcc but realized how limited it was. I tried making my own component but I am not a JS developer. This was exactly what I was looking for though, so great job, and thanks for publishing it @lewkoo !

Now onto the actual discussion (and please let me know if this isn't the right place for this) - I have gone through a few iterations of my app and Dashvis has worked perfectly up until adding more dash pages. My app takes user input and returns a very simple DashNetwork element:

Transaction_Network.py
```python

dash.register_page(__name__, path='/transaction_network', title='Transaction Network')

def layout():
    return html.Div([
        dbc.Col([
            dmc.LoadingOverlay([html.Div(id='networkContainer', style={'height': '100%'}, className="border round border border-light")], style={"height": "100%"}),
            ], style={"height": "100%"}),

@callback(
    Output('networkContainer', 'children'),
    Input('generate_network_btn', 'n_clicks'),
    State('block_data_store', 'data'),
    prevent_initial_call=True
)
def create_network(clicks, block_dict, settings):
    if block_dict is None or clicks == 0 or clicks == None:
       PreventUpdate
  

    block_instance = Network.BlockData.from_dict(block_dict)
    #this is the creation of a data class used elsewhere in my code

    if block_instance:
        network_data = Backend.create_network(block_instance) 
        #this is the actual creation of the nodes and edges JSON data

        network = DashNetwork(
            id='network',
            style={'height': '100%'},
            options=networkOptions,
            enableHciEvents=[
                'selectNode', 'selectEdge', 'deselectNode', 'deselectEdge', 'hoverNode', 'hoverEdge', 'showPopup', 'hidePopup'
            ],
            enablePhysicsEvents=False,
            enableOtherEvents=True,
            data = network_data
        )


        return network
``` 

Initially, I thought it was an issue with my Pages implementation but my page navigation works fine when I don't create the network visual. However after the network is rendered, and I attempt to navigate to another page and the content does not appear, and I get these errors: 

![image](https://github.com/lewkoo/dashvis/assets/103399058/a6730201-f2c9-45d0-b101-206022764fbe)
(the first 7 errors are all the same and provide no more info beyond "Cannot read properties of undefined (reading 'type')")
![image](https://github.com/lewkoo/dashvis/assets/103399058/69900968-c9e8-41be-a8de-1d165a1521e5)

At this point, I am pretty confused. My Dash Pages implementation seems fine and my Transaction_Network.py file has a simple implementation. I have tried different page/navigation implementations but I continue to get these errors. So I'm not really sure if this is a question for Dash or Dashvis. I did find [this thread in the Plotly Community Forum](https://community.plotly.com/t/error-with-navlink/72155/2) with the same errors, but no insight into the fix. 

I think that this is an issue with my implementation, but I really do not know. I'm thinking of using some hidden div magic to avoid the error, but that won't actually fix it. As I mentioned, this is my first 'side project' and I am not a professional developer so any insight would be helpful! Hopefully, this can help others in the future as well.

**Comment by lewkoo** [comment]
Hi @McGinn1014! Thank you for your words regarding `dashvis`. I am not a JS developer myself - I needed to create it for something else I am developing and decided to open source it. 

Please note that I am basing my thoughts on what you've posted above and, as I have no access to the rest of your code, my opinion might be totally wrong. 

I haven't ever used Dash multi-page myself but I suspect it has nothing to do with your error. Instead, I think it is sourced in your `networkOptions` dictionary that you're passing to the `options` parameter when calling `DashNetwork` constructor. `options` parameter has a very complex structure - yet, it was easier to just exactly mirror the structure as-is on `vis.js` than to create my own and necessitate a creation of a separate documentation and additional complexity. 

Looking at vis.js docs, I see that `resize` is not a valid attribute of `options` dictionary but is instead published as a callback property. So I would advice you to double check your `networkOptions` dictionary and remove the `resize` key. 

Let me know if that helped.

**Comment by McGinn1014** [comment]
Hi @lewkoo, unfortunately that doesn't work as I do not have a resize key in my `networkOptions`, here is the dictionary for reference:
```python
networkOptions = {
            'nodes': {
                'shape': 'circle',
                'color': {
                    'border': '#2B7CE9', 
                    'background': '#97C2FC', 
                    'highlight': {
                    'border': '#2B7CE9',
                    'background': '#D2E5FF',
                    },
                },
                'font': {
                    'size': 14, 
                },
                
            },
            'edges': {
                'width': 2, 
                'color': '#848484', 
                'arrows': {
                    'middle': {
                        'enabled': True
                    }
                },
                'arrowStrikethrough': True,
                            
            },
            'physics': {
                'enabled': True, 
                'solver': 'forceAtlas2Based',
                'stabilization': {
                    'enabled': False, 
                    'fit': False
                }
            },

            'interaction': {
                'hover': True,
                'selectConnectedEdges': False,
                'dragNodes': False,
            },
            'layout': {
                'improvedLayout': False
            },
            'manipulation': {
                'enabled': False
            }
        }
```

**Reply by lewkoo** [reply]
Is it possible to see more detailed logs?

**Reply by McGinn1014** [reply]
I've recreated the issue in a more simple app. Below is the full code, note that the navigation buttons work fine until pressing the `show_network` button. The console error logs are below as well.

Test_App.py
```python
import dash
from dash import Input, Output, State, dcc, html, _pages, no_update, callback
from dash.exceptions import PreventUpdate
from dashvis import DashNetwork
import dashvis.stylesheets
import dash_bootstrap_components as dbc



app = dash.Dash(__name__, use_pages=True, external_stylesheets=[dbc.themes.DARKLY, dashvis.stylesheets.VIS_NETWORK_STYLESHEET])

app.layout = html.Div([
    html.H2("Dashvis bug test"),
    dcc.Link("Home", href="/"),
    dcc.Link("Network", href="/network"),
    
        
    dash.page_container
])

if __name__ == '__main__':
    app.run(debug=True)
```

home.py
```python
import dash
from dash import Input, Output, State, dcc, html, _pages, no_update, callback
from dash.exceptions import PreventUpdate
from dashvis import DashNetwork
import dashvis.stylesheets
import dash_bootstrap_components as dbc

dash.register_page(__name__, path="/")

def layout():
    return html.Div([
        html.H4('this is a home page')
    ])
```

network.py
```python
import dash
from dash import Input, Output, State, dcc, html, _pages, no_update, callback
from dash.exceptions import PreventUpdate
from dashvis import DashNetwork
import dashvis.stylesheets
import dash_bootstrap_components as dbc

networkOptions = {
            'nodes': {
                'shape': 'circle',
                'color': {
                    'border': '#2B7CE9', 
                    'background': '#97C2FC', 
                    'highlight': {
                    'border': '#2B7CE9',
                    'background': '#D2E5FF',
                    },
                },
                'font': {
                    'size': 14, 
                },
                
            },
            'edges': {
                'width': 2, 
                'color': '#848484', 
                'arrows': {
                    'middle': {
                        'enabled': True
                    }
                },
                'arrowStrikethrough': True,
                            
            },
            'physics': {
                'enabled': True, 
                'solver': 'forceAtlas2Based',
                'stabilization': {
                    'enabled': False, 
                    'fit': False
                }
            },

            'interaction': {
                'hover': True,
                'selectConnectedEdges': False,
                'dragNodes': False,
            },
            'layout': {
                'improvedLayout': False
            },
            'manipulation': {
                'enabled': False
            }
        }



dash.register_page(__name__, path="/network")

def layout(): 
    return html.Div([
        html.H3("Dashvis Network"),
        html.Button("show_network", id="network_button"),
        html.Div(id='network_container')
    ])

@callback(
    Output('network_container', 'children'),
    Input('network_button', 'n_clicks'),
    prevent_initial_call=True
)
def show_net(clicks):
    if clicks is None or clicks == 0:
        dash.no_update
    network = DashNetwork(
            id='test_network',
            style={'height': '100%'},
            options=networkOptions,
            enableHciEvents=[
                'selectNode', 'selectEdge', 'deselectNode', 'deselectEdge', 'hoverNode', 'hoverEdge', 'showPopup', 'hidePopup'
            ],
            
            enablePhysicsEvents=False,
            enableOtherEvents=True,
        )
    return network
```

Here are screenshots of the console logs: 

![image](https://github.com/lewkoo/dashvis/assets/103399058/b51d717d-31c5-4650-b8b6-4bc671f50f69)

![image](https://github.com/lewkoo/dashvis/assets/103399058/787a4f04-430c-4ea8-abaa-0a7443de9fd5)

![image](https://github.com/lewkoo/dashvis/assets/103399058/57bb2525-6449-4a6e-aa4f-ea0e9e7ea80e)

![image](https://github.com/lewkoo/dashvis/assets/103399058/5104b177-1216-41e1-b6ab-ffee70720b16)

**Reply by McGinn1014** [reply]
So in my sample app, setting `enableOtherEvents=False` solves the error, so when you referenced the `resize` key in my network options, it was actually the resize event causing the issue. Not sure how to fix that. 

Keeping `enablePhysicsEvents=True` also works fine, but if I attempt to navigate away too quickly/before the network has stabilized, I get a similar error as above: 
![image](https://github.com/lewkoo/dashvis/assets/103399058/81ceb21a-9cda-40c1-bb6b-134783f58348)

**Comment by lewkoo** [comment]
First of all, I had to change 

```
    if clicks is None or clicks == 0:
        dash.no_update
```

to 

```
    if clicks is None or clicks == 0:
        raise dash.exceptions.PreventUpdate
```

to run your code. 

WIth that out of the way, I think I understood what's wrong. When either `enableHciEvents`, `enablePhysicsEvents` or `enableOtherEvents` is not equal to exactly False, `dashvis` will add a event listener to the network to update certain properties of the network as they change and report these changes to dash (so that you can use them in a callback, for example). 

For the rest of the discussion, see issue #14 and please upgrade to the latest version - 0.1.9.
